### DISCUSSION TITLE
Helm Chart

### DISCUSSION THREAD

**Discussion by mmuenker** [advplyr/audiobookshelf#2851] [Ideas]
Having a helm chart for easy deployment on a Kubernetes cluster would be convenient. I am interested in contributing to this project. The first step is determining how to get started. 

- (A) Would it be best to create a new GitHub repository? 
- (B) Or should I open a merge request instead? 
- (C) Additionally, we need to figure out how to handle ownership. 
  - (C.I) Should I create and register the helm chart myself, 
  - (C.II) or should a core maintainer take care of that?

**Comment by nichwall** [comment]
A) The best place for it would be adding to the install section in the website repository. Someone shared a helm chart in the issues in the repo, but no idea if it's up to date/works still
https://github.com/audiobookshelf/audiobookshelf-web

B) A fork then PR to the documentation works well

C) Not sure how helm chart ownership works, is it like a docker compose file when it's just a text file, or is it like an install binary?

**Reply by mmuenker** [reply]
A) I will examine the current helm chart, and at first glance, it appears to be a functional version.
B) okay
C) It is similar to Docker, where a registry/repository is required to store bundled helm charts in a zip format. The user then needs to add this registry to their local helm client in order to install the chart. This process is comparable to adding a second Docker registry. There are various methods available for achieving this, such as using a [GitHub repository](https://github.com/helm/chart-releaser) as a registry or utilizing a dedicated section for helm charts within a [GitLab project](https://docs.gitlab.com/ee/user/packages/helm_repository/). Ultimately, a helm repository is simply a webserver that has a specific file structure. - https://helm.sh/docs/topics/chart_repository/.

**Reply by nichwall** [reply]
Ah, cool! I think it could fit as a repository under the Audiobookshelf organization, as long as there was a disclaimer that it is community maintained. Looking at those helm charts that were shared in the other issue, it looks like the only thing that would need to change would be updating the release number?

- Would anything else need to change for releases, and how would users be able to run a different version if they want to? We want people to be staying up to date, but some people choose to run older versions due to breaking changes or performance issues.
- Looks like the helm charts are pulling the latest code from this repo, or do they pull from the latest release? Are they building from source?

If it's only the version number that changes, that could be handled by a GH action.

**Reply by mmuenker** [reply]
> Would anything else need to change for releases, and how would users be able to run a different version if they want to? We want people to be staying up to date, but some people choose to run older versions due to breaking changes or performance issues.

This can be easily configured in the config file for the helm deployment. This could look like this:

**values.yaml**
```yaml
image:
  name: ghcr.io/advplyr/audiobookshelf
  tag: v1.3.4
```

> Looks like the helm charts are pulling the latest code from this repo, or do they pull from the latest release? Are they building from source?

The helm chart describes the deployment of a (docker) container in Kubernetes. Think of helm charts as Docker Compose for Kubernetes.

**Reply by mmuenker** [reply]
@nichwall @advplyr I have created a repository with a helm chart. 

Since I am not well-versed in github actions, I have created a gitlab project for the CI/CD process. Currently, the CI/CD configuration checks each night for a new version. If one is found, it updates the default app version to the latest one and creates a nightly build for the helm chart. Additionally, the gitlab repository is mirrored with a github repository.

- https://gitlab.com/mmuenker/audiobookshelf-chart
- https://github.com/mmuenker/audiobookshelf-chart
